src/Decompiler/Typing/ExpressionTypeAscender.cs

```diff
@@ -147,8 +147,11 @@ public DataType VisitBinaryExpression(BinaryExpression binExp)
             }
             else
                 throw NYI(binExp);
-            binExp.TypeVariable.DataType = dt;
-            binExp.TypeVariable.OriginalDataType = dt;
+            if (binExp.TypeVariable != null)
+            {
```

Comment:
I'm a concerned about these null checks. Setting the `TypeVariable` property is done as part of the "core" type analysis which happens much later in the current process, but here obviously we want to do it almost immediately after SSA transformation. 

More and more, it seems to me that `Expression` should lose the `TypeVariable` property and move it into the `TypeStore`. Then these statements could be a simple:
`typestore.SetDataTypes(binExp, dt, dt);` and you wouldn't need the the null tests. 
It's tough to reason about this when I'm on vacation though -- I will have to think about this some more on my way back to Stockholm.